### DISCUSSION TITLE
Changing the log level of shaded gRPC classes in glowroot agent

### DISCUSSION THREAD

**Discussion by ingmars1709** [glowroot/glowroot#1235] [Q&A] [answered]
I want to suppress some logging about expected connection errors.

How can I change the log level for the class: `org.glowroot.agent.shaded.io.grpc.internal.ManagedChannelImpl` to the `INFO` level?

It seems it doesn't use Logback, but instead uses `java.util.logging.Logger`
https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

I've tried running with: 
`java -javaagent:./<path-to>/glowroot.jar -Djava.util.logging.config.file=./<path-to>/logging.properties`

And then in `logging.properties` tried the following config lines:

```
org.glowroot.agent.shaded.io.grpc.internal.ManagedChannelImpl.level = INFO
io.grpc.internal.ManagedChannelImpl.level = INFO
```

But it isn't being picked up and the logging is still done on `WARN` level.

The log message I want to suppress:

```
2024-05-14 16:03:35.304 WARN  o.g.a.s.i.g.i.ManagedChannelImpl - [Channel<1>: (glowroot-monitoring-collector:8181)] Failed to resolve name. status=Status{code=UNAVAILABLE, description=Unable to resolve host glowroot-monitoring-collector, cause=java.lang.RuntimeException: java.net.UnknownHostException: glowroot-monitoring-collector: nodename nor servname provided, or not known
	at org.glowroot.agent.shaded.io.grpc.internal.DnsNameResolver.resolveAddresses(DnsNameResolver.java:223)
	at org.glowroot.agent.shaded.io.grpc.internal.DnsNameResolver.doResolve(DnsNameResolver.java:282)
	at org.glowroot.agent.shaded.io.grpc.internal.DnsNameResolver$Resolve.run(DnsNameResolver.java:318)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.net.UnknownHostException: glowroot-monitoring-collector: nodename nor servname provided, or not known
```

**Comment by Never-lab** [comment, marked as answer]
## Technical analysis

`-Djava.util.logging.config.file` won't work here. During shading, `java.util.logging` is relocated to `org.glowroot.agent.jul`, which bridges straight to SLF4J/Logback — and `setLevel()` on that bridge is a no-op (`agent/core/.../jul/Logger.java`). So `logging.properties` is never consulted for gRPC output.

Glowroot agent logging is controlled by **Logback** via `glowroot.logback.xml` (shaded inside the JAR; override by copying it next to `glowroot.jar` in the conf dir — see the header comment in the bundled file and `MainEntryPoint.initLogging`).

The default config already sets `org.glowroot.agent.shaded` to `warn`, which is why you still see these messages — JUL `WARNING` maps to Logback `WARN`, and a `warn` threshold includes them.

**To suppress** the DNS-resolution noise, add to your override `glowroot.logback.xml`:

```xml
<logger name="org.glowroot.agent.shaded.io.grpc.internal.ManagedChannelImpl" level="error" />
```

(Use `off` to silence entirely. Setting `info` won't help — in Logback, `info` still emits `warn`.)

The underlying `UnknownHostException` for `glowroot-monitoring-collector` is expected when Central isn't reachable yet; this is just noisy logging, not a functional error.

---

*Community analysis; not an official maintainer statement.*

**Comment by Never-lab** [comment]
Closing — this is a **configuration how-to**, not a bug.

Per [CONTRIBUTING.md](https://github.com/glowroot/glowroot/blob/main/CONTRIBUTING.md) and [README](https://github.com/glowroot/glowroot#community), questions like this belong in **GitHub Discussions → Q&A**, not Issues.

Answer and context moved to: https://github.com/glowroot/glowroot/discussions/1227

**TL;DR:** use a `glowroot.logback.xml` override, not `logging.properties`.
